### PR TITLE
NOCK fixtures

### PR DESCRIPTION
**Pull request by cryptomail**
## Pull Request Description

Adds nock fixtures to tests.
https://github.com/blakmatrix/node-zendesk/issues/361

---

## Related Issue(s)

- [x] This PR fixes/closes issue #361 
- [ ] This is a new feature and does not have an associated issue.

---

## Additional Information

- [ ] This change is a breaking change (may require a major version update)
- [ ] This change is a new feature (non-breaking change which adds functionality)
- [x] This change improves the code (e.g., refactoring, etc.)
- [x] This change includes dependency updates

---

## Test Cases

Include any test cases or steps you took to test your changes. If you have added new functionality, please include relevant unit tests.

---

## Documentation

- [x] I have updated the documentation accordingly.
- [x] No updates are required.

---

## Checklist

- [x] I have read the [CONTRIBUTING](./CONTRIBUTING.md) documentation.
- [x] My code follows the coding standards of this project.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/client/helpers.js [hunk 1]
```diff
@@ -12,6 +12,45 @@ const failCodes = {
...
+ *
+ * @param {Object} response - Response object.
+ * @returns {Object} - Target header object.
+ */
+function responseBodyJSON(response) {
+  if(['development', 'test'].includes(process.env.NODE_ENV)) {
+    return JSON.parse(response.body);
+  }
+  return response.json();
+}
```
Comment (ecoologic) [inline src/client/helpers.js:53]: 🤔  This breaks production development parity. Why is this required? How about parsing the body in all env? (line 50)

File: package.json [hunk 2]
```diff
@@ -24,8 +24,8 @@
   },
   "main": "./src/index.js",
   "scripts": {
-    "test": "vitest run",
-    "test:coverage": "vitest run --coverage",
+    "test": "NODE_ENV=development vitest run",
+    "test:coverage": "NODE_ENV=development vitest run --coverage",
```
Comment (ecoologic) [inline package.json:28]: This is confusing to me.

File: src/client/transporter.js [hunk 3]
```diff
@@ -53,10 +58,11 @@ class Transporter {
...
     let result = {};
     if (
-      response.status !== 204 &&
-      response.headers.get('content-type')?.includes('application/json')
+
+      responseGetStatus(response) !== 204 &&
+        responseHeaderRetrieve(response, 'content-type')?.includes('application/json')
     ) {
-      result = await response.json();
+      result = await responseBodyJSON(response);
```
Comment (ecoologic) [inline src/client/transporter.js:65]: I wonder if the the above check is enough to ensure there are no parsing errors in this function. ie: Do we miss a `catch`?

**Review by ecoologic — COMMENTED** [review]
Looks like I miss some context, I don't know nock fixtures, I used https://mswjs.io which I loved and I never had to write these workaround functions I see below.

File: src/client/helpers.js (hunk 1 above)
Comment (cryptomail) [reply src/client/helpers.js:53]: Nock  fixtures only work with request lib.  Sell all the machinations necessary to employ request instead of native request in node.js

Comment (cryptomail) [reply src/client/helpers.js:53]: Is there a transparent record mode / playback mode in msw?
IE logically:
record(fixturname) {
if fixture found use existing fixture
else
     allow request to go through
do work/test
}

File: src/client/transporter.js [hunk 4]
```diff
@@ -3,7 +3,12 @@
 const {AuthorizationHandler} = require('./authorization-handler');
 const {CustomEventTarget} = require('./custom-event-target');
 const {EndpointChecker} = require('./endpoint-checker');
-const {assembleUrl} = require('./helpers');
+const {assembleUrl, responseHeaderRetrieve, responseBodyJSON, responseGetStatus} = require('./helpers');
+
+let requestApi = null;
+if(['development', 'test'].includes(process.env.NODE_ENV)) {
```
Comment (ecoologic) [inline src/client/transporter.js:9]: given the amount of duplication, maybe we can extract this in its own method. This could be a const set at startup.

I'm also unsure why we need the `test` env in here if we're forcing everything to be development.

```suggestion
if(MOCKED_API) {
```

Comment (cryptomail) [reply src/client/transporter.js:9]: Running through intellij yielded development as default

File: src/client/helpers.js (hunk 1 above)
Comment (ecoologic) [reply src/client/helpers.js:53]: I might be wrong and I admit I don't know the library, but I find it strange that a production ready library requires us to manually write all this untestable code.

Comment (ecoologic) [reply src/client/helpers.js:53]: > Is there a transparent record mode / playback mode in msw?

yes, it's like the VCR gem. It's been 4 years that I don't touch it, but I remember being very happy using it.

Comment (cryptomail) [reply src/client/helpers.js:53]: > I might be wrong and I admit I don't know the library, but I find it strange that a production ready library requires us to manually write all this untestable code.

Please help by integrating msw, and rip out the conditional code, keeping the above `notional` form factor of
record() {
if fixture found use fixture
else allow request to flow through and capture

do work/test
end record
}

Comment (cryptomail) [reply src/client/helpers.js:53]: Node version I use is 18.4

Comment (ecoologic) [reply src/client/helpers.js:53]: 💣  - My memory failed me - MSW doesn't allow you to record missing calls automatically. You need to do that manually 😞

**Review by blakmatrix — CHANGES_REQUESTED** [review]
Hi @cryptomail,

First off, thank you for the efforts you've put into enhancing the test coverage. I appreciate your commitment to improving the quality of `node-zendesk`.

Given the concerns about the use of `process.env.NODE_ENV` checks in the main library and the dependency on the deprecated `request` library, I've made some updates. I've introduced a way for the library to be HTTP Client Agnostic. This means the main code within `src` doesn't have to be altered based on the testing environment or the HTTP client used. You can find the changes here: [link to the commit](https://github.com/blakmatrix/node-zendesk/commit/9b4489a3a18e5c0a3732ac56b246cab5ae4e5ff7).

I'd like to request a few things from your side:

1. Please update the tests to use the [new approach](https://github.com/blakmatrix/node-zendesk/commit/9b4489a3a18e5c0a3732ac56b246cab5ae4e5ff7#diff-84701e9f6222583b5e5f7da649e5b0f4db307648c8e009bf9a49599526cb9fb3), integrating `request` and `nock`.
2. It would be great if you could provide some documentation or comments in the test suite explaining the integration and usage of `request` and `nock`. This would be beneficial for anyone looking into the tests in the future.
3. I have a slight concern about potential drift: when testing with `request` but internally allowing users to opt for different clients like `fetch`, `axios`, etc. Do you foresee any issues arising from this difference in testing and potential usage scenarios?

Once again, thanks for your contribution, and I'm looking forward to your updates on this!

Best,
Farrin

**Comment by cryptomail** [comment]
I'll do my level-best to work through your changes.  Thanks for the suggestions.

**Comment by cryptomail** [comment]
OK keeping this for reference, but this is going to be closed shortly.  Starting a new PR shortly.

**Comment by cryptomail** [comment]
Also, I do foresee some potential risks in using nock/request as it's not the production code path.  Until node.js gets its act together along with Nock, I don't see a way out, and I see that as a risk we **should** be able to assume.  Given how "pluggable" request libraries are (but not pluggable in the sense of being able to wire capture properly for nock....) I feel the end result is perfectly fine and reasonable.
The social commentary I have is this: A well established project like Nock used by 1000's? of customers was run over by Node.js.  Whether in the name of progress or not, this is where we are.  If I were directing things(excuse a bit of my megalomaniacal stance here), I would make Node.js recognize the widely accepted and useful Nock library and get in there and HELP Nock transform their library to be as compatible as possible to provide as little disruption to users as possible.  🪑 end rant.

**Comment by cryptomail** [comment]
superceded by #375 
#375 employs new internal factorization for transport.
